### ISSUE TITLE
REST_GetRecords not working for me.

### ISSUE THREAD

**Issue by apunekar** [fourctv/JS44D#2]
```
const body = {'query': ['PERSON.Last Name;contains;ab']};
    this.fourD.call4DRESTMethod('REST_GetRecords', body)
    .subscribe(
        response => {
          console.log('error:' + response.text());
        },
        error => {
          console.log('error:' + error.text());
        });
```

Above is the code I am using to get records from PERSON table with last name conatins 'ab'. I am getting 500 Error. I am able to use the fourD.signIn. after this I want to call this method. Its not workin

**Comment by fourctv** [comment]
@apunekar Is there really a space on that "Last Name" field? Because that is a problem, and is not supported...

Although 4D does support spaces in field names, that is definitely not recommended.

So, either your code is incorrect, or you have to change that field's name.

**Comment by amit-punekar-membermax** [comment]
@fourctv I modified my code to look something like this

```
const body = {'tableName': 'PERSON',
                  'Columns': [
                  {'name': 'PERSON ID', 'field': 'PERSON.PERSON ID'},
                  {'name': 'First Name', 'field': 'PERSON.FIRST NAME'},
                  {'name': 'Last Name', 'field': 'PERSON.LAST NAME'}],
                  'QueryString': ['PERSON.Last Name;contains;ab']};
    this.fourD.call4DRESTMethod('REST_GetRecords', body)
    .subscribe(
        response => {
          console.log('error:' + response.text());
        },
        error => {
            console.log('error:' + error.text());
        });
```
the columns parameter is not working here 
`$fieldList:=JSON Parse(BLOB to text($columnsList;UTF8 text without length))`

fieldList is retuning nil over here....

Regarding the space in the Column Name.. Yes there are spaces in the column name [PERSON]LAST NAME and many moer such columns with space in them.

**Comment by fourctv** [comment]
why are you using 'call4DRESTMethod' to get records? You should be using a FourDModel instance.
You should create a 'PERSON' FourDModel and then use something like:
`
let rec:Person = new Person()
rec.getRecords(query,columns).then(recList => {
...
}
`

**Comment by fourctv** [comment]
Take a look at the wiki here:

https://github.com/fourctv/JS44D/wiki/Data-Modelling,-How-to-Generate-a-Model

**Comment by apunekar** [comment]
> Take a look at the wiki here:
> 
> https://github.com/fourctv/JS44D/wiki/Data-Modelling,-How-to-Generate-a-Model

Yes I have already generated a model for PERSON table. Ok I will chec kthe solution given by you.. So what do you advise .. when to use  'call4DRESTMethod'?

**Comment by fourctv** [comment]
"call4DRESTMethod" is to be used to call your own REST entry points. All the REST entry points in the RESTAPI are covered by methods and functions in this JS44D library.
So better to use the proper functions in JS44D, then calling "call4DRESTMethod" manually.

**Comment by apunekar** [comment]
@fourctv Thanks a lot. Just to go back to your point about having spaces in the field names of the 4D tables I cam acorss a discussion in which we can it by adding single quotes to that field. here is the link http://forums.4d.com/Post/EN/25635039/1
I tried it too and its working for me too. Now if you can fix it in your routine RESTParseQuery to add single quotes to all the field and then query using the ds.query() it would fix most of the problems as we have many fields with space in them.

![Screenshot 2021-01-29 165641](https://user-images.githubusercontent.com/67050931/106271620-cd3cfb00-6255-11eb-909b-82f0dcaf91b9.jpg)

**Comment by fourctv** [comment]
Sorry @apunekar but allowing spaces in table or field names is against any naming conventions, on any language or development environment. That is a legacy feature of 4D, which they had to keep for compatibility reasons. It is not even supported by 4D's own Wakanda environment.
So, my recommendation is that you eliminate those spaces on all your field names. Keeping that will only cause you trouble in the future.

**Comment by amit-punekar-membermax** [comment]
@fourctv Thanks Understood. Thanks for the help anyways. Will look forward and change the field names.

**Comment by fourctv** [comment]
@apunekar glad to be of help. Let me know if you need anything else.

**Comment by amit-punekar-membermax** [comment]
@fourctv  I am try to use the getRecord() and passing the record number as a parameter.
this.fourDModel.getRecord(recordnumber)
  .then(recList => {
        console.log(recList);
        },
      error => {
        console.log(error);
      });
  }
  It calls the REST_LoadData on the 4D side but checks for table name and returns no records because of table name not found.

**Comment by fourctv** [comment]
@apunekar you do not use FourDModel class directly. You have to extend FourDModel with your own table classes, and then use them to get/set data from 4D.
See: https://github.com/fourctv/JS44D/wiki/Data-Modelling,-How-to-Generate-a-Model, which explains how you generate your own table classes, which extend FourDModel.

Then you could have something like this, assuming you have a People table:

> let my rec:People = new People();
rec.getRecord(recordnumber)
.then(rec => {
... deal with your record
}

**Comment by amit-punekar-membermax** [comment]
@fourctv   Perfect. Thanks. and what exaactky does this populateModelData() do.
I have called in such way. Is this right ?
const rec: PERSON = new PERSON();
    const query = {'query': ['PERSON.Last Name;begins with;ab']};
    const columns: string[] = ['PERSON.First Name', 'PERSON.Last Name', 'PERSON.Mail Address1', 'PERSON.Mail City', 'PERSON.Mail State', 'PERSON.Mail Zip'];

    rec.getRecords(query, columns)
    .then(recList => {
        console.log(recList);
        rec.populateModelData(recList.models);
        this.dataSource = new MatTableDataSource(recList.models);
        this.dataSource.paginator = this.paginator;
        },
      error => {
        console.log(error);
      });

**Comment by fourctv** [comment]
populateModelData() is used to populate a FourDModel class with attributes from another object.

There are cases where you cannot use a FourDModel sub-class. Where you need a 'plain' object. Then you can use extractModelData(), which will do just that generate a 'plain' object whose attributes come from a FourDModel object.

populateModelData() does the opposite.

And FYI, getRecords() returns a FourDCollection. In your example, 'recList' is a FourDCollection instance, whose "models" attribute gives you an array of FourDModel sub-classes (records) retrieved from 4D. (as you've probably seen on your console.log().

Thus, that populateModelData() has no place in your code example...

**Comment by amit-punekar-membermax** [comment]
@fourctv  Thanks fot the explanation. Another issue which I am facing is with the authentication object returned from REST_Authenticate called from signIn(). The authenticate.json file generated has the 'valid' but in the console I can't see the valid property. Please find the attached screenshot.
![Screenshot 2021-02-04 153525](https://user-images.githubusercontent.com/71878462/106877455-c6ebca80-66fe-11eb-9dad-b6941305a9ec.jpg)

**Comment by fourctv** [comment]
If you are looking at the FourDInterface 'authentication' property it does not have that 'valid' attribute.
JS44D sign process checks that 'valid' attribute and if 'true', it accepts the sign in and saves the 'session' attribute into that 'authentication' object property.
There is no need to save 'valid', it's only use is in the sign in process to know the user/pwd use for authentication are indeed valid.
If REST_Authenticate returns 'valid:false', then the sign in process does not complete. It sticks on the sign in dialog, waiting for a valid user/pwd.

**Comment by apunekar** [comment]
Thanks. As you said in your documentation that we need to send session key and hash in every request except Authenticate and Application Version. I have implmented a interceptor for the same. How can I call four method using interceptor. For eg how can I call getRecords() using a interceptor OR is there any other way too add session key and hash in the header.

**Comment by fourctv** [comment]
What do you mean by 'interceptor'? 
If you want to call your own HTTP entry point, you should take a look at the [FourDInterface "call4DRESTMethod" ](https://github.com/fourctv/JS44D/wiki/FourDInterface-Class), that method will add a session key and hash code to your URL+arguments, before sending it to 4D.

**Comment by apunekar** [comment]
Yes exactly, By interecptor  I meant http entry point and to add headers to the http request. So what you mean to say is that we do not need to add session key and hash  to all the methods inside the FourDModel class, by default the session key and hash is added in the header. If we need to make calls to our custom REST API methods to 4D server, then we need to call it through  FourDInterface "call4DRESTMethod" method which will add a session key and hash code. Is that correct?

**Comment by fourctv** [comment]
@apunekar that is correct.
The basic way to do it is to use something like:
`constructor (private fourD:FourDInterface) {
}`

Then when you want to call your REST entry point you do something like:
`
this.fourD.call4DRESTMethod("myrestmethod";{parm1:1, parm2:"2", ...}).then(response => {...});
`

**Comment by apunekar** [comment]
@fourctv Thanks a lot.

**Comment by apunekar** [comment]
@fourctv Guess what I am back again....LOL
I am facing couple of issues 
First with getRecords(). I am sending a query property as param in getRecords(). But getting error if I add more than one condition. In this example AND 
Secondly, I was trying a record using getRecord() and sending the query property in the parameter but it gives me error.
Please check both the isssues in the attached screenshot.
![Untitled](https://user-images.githubusercontent.com/67050931/107780542-cee5e300-6d6c-11eb-8b8a-b4d9a850e4ee.png)

**Comment by apunekar** [comment]
> @fourctv Guess what I am back again....LOL
> I am facing couple of issues
> First with getRecords(). I am sending a query property as param in getRecords(). But getting error if I add more than one condition. In this example AND
> Secondly, I was trying a record using getRecord() and sending the query property in the parameter but it gives me error.
> Please check both the isssues in the attached screenshot.
> ![Untitled](https://user-images.githubusercontent.com/67050931/107780542-cee5e300-6d6c-11eb-8b8a-b4d9a850e4ee.png)

@fourctv My Bad, Fixed my code to resolve issue# 1. But issue#2 still exists.

**Comment by fourctv** [comment]
@apunekar check out the getRecord() parameters (https://github.com/fourctv/JS44D/wiki/FourDModel-Class)

That function takes 3 arguments getRecord(record number, record ID, query).

So in your case you should do:

`
this.userRecord.getRecord(null, null, query)
`

Give that a try...

Also the query parameter should be:

`
const query = {query:['User.Username;=;'+userName+';AND', 'User.UserPassword;=;'+password+';AND'
`

The 'query' property is an array of 4D query arguments. That array gets translated into a sequence of QUERY statements on 4D side.

**Comment by apunekar** [comment]
@fourctv Already had fixed the query parameter structure. Thanks for getRecord(record number, record ID, query) clarification.

**Comment by apunekar** [comment]
@fourctv Trying to create a query for join table this is how I have created the query object
const query = {join : [{joinTable: 'USER', joinFK: 'USER.UserID', joinPK: 'StaffRole.StaffID'}]};
 But no luck. Its not working. Need help

**Comment by apunekar** [comment]
> {join : [{joinTable: 'USER', joinFK: 'USER.UserID', joinPK: 'StaffRole.StaffID'}]};

Modified it to this
const query = {joinTable: 'USER', joinFK: 'USER.UserID', joinPK: 'StaffRole.StaffID', join: []};
Is this correct ?

**Comment by apunekar** [comment]
@fourctv  
A Join Query is not working for me. Whatever I pass in the Join parameter gets blanked out when the RESTRunQuery method is recalled. In RESTRunQuery when parameter $2 is assigned to variable $queryString, the variable gets values assigned as blank string. 
const query = {joinTable: 'USER', joinFK: 'USER.UserID', joinPK: 'StaffRole.StaffID',
                 join: ['all']};
